pyteal/ast/abi/type.py

```diff
@@ -231,7 +240,22 @@ def store_into(self, output: BaseType) -> Expr:
                 f"ABI return subroutine deferred_expr is expected to be typed {output.type_spec().storage_type()}, "
                 f"but has type {declaration.deferred_expr.type_of()}."
             )
-        return output.stored_value.slot.store(self.computation)
+
+        validate_in_store: bool = True
+
+        if (
+            isinstance(self.computation, SubroutineCall)
+            and self.computation.output_kwarg
+        ):
+            assert types_match(
+                self.computation.output_kwarg.abi_type.storage_type(),
+                output._stored_value.storage_type(),
+            )
+            validate_in_store = False
```

Comment:
What you are point at is the reason that `ABIReturnSubroutine` evaluates to `TealType.none`, but recall the syntax of `ABIReturnSubroutine`, which returns an `Expr` that:
- either has type `void`, which is `TealType.none`
- or does have an output ABI value, but the `Expr` evaluates to `TealType.none`
  - in scratch slot based compilation, we use deferred expr to load to stack
  - in frame based compilation, we do nothing for result is already placed on stack

  and then we wrap the computation with `ReturnedValue`.

In either case, I don't feel right about changing this line...

Reply:
After some more thoughts, I think what determines here is the scratch slot behavior, that takes the encoding off the stack.

I would propose to let return type to be concrete, only when the return value is in the context of frame pointer, rather than being stored in a scratch slot.

Reply:
resolved by introducing change of `type_of` in `SubrouineCall` object, removing all of `validate_type` argument.